types/src/v0/impls/state.rs

```diff
@@ -253,130 +303,329 @@ impl ValidatedState {
+        Ok(())
     }
+    /// Validate that proposal block size does not exceed configured
+    /// `ChainConfig.max_block_size`.
+    fn validate_block_size(&self) -> Result<(), ProposalValidationError> {
+        let block_size = self.proposal.block_size as u64;
+        if block_size > *self.expected_chain_config.max_block_size {
+            return Err(ProposalValidationError::MaxBlockSizeExceeded {
+                max_block_size: self.expected_chain_config.max_block_size,
+                block_size: block_size.into(),
+            });
+        }
+        Ok(())
+    }
+    /// Validate that [`FeeAmount`] (or sum of fees for Marketplace Version) is
+    /// sufficient for block size.
+    fn validate_fee(&self) -> Result<(), ProposalValidationError> {
+        // TODO this should be updated to `base_fee * bundle_size` when we have
+        // VID per bundle or namespace.
+        let Some(amount) = self.proposal.header.fee_info().amount() else {
+            return Err(ProposalValidationError::SomeFeeAmountOutOfRange);
+        };
 
-    // Validate timestamp increasing.
-    // TODO add test https://github.com/EspressoSystems/espresso-sequencer/issues/2100
-    if proposal.timestamp() < parent_header.timestamp() {
-        return Err(ProposalValidationError::InvalidTimestampNonIncrementing {
-            proposal_timestamp: proposal.timestamp(),
-            parent_timestamp: parent_header.timestamp(),
-        });
+        if amount < self.expected_chain_config.base_fee * self.proposal.block_size {
+            return Err(ProposalValidationError::InsufficientFee {
+                max_block_size: self.expected_chain_config.max_block_size,
+                base_fee: self.expected_chain_config.base_fee,
+                proposed_fee: amount,
+            });
+        }
+        Ok(())
+    }
+    /// Validate that proposal height is `parent_height + 1`.
+    fn validate_height(&self) -> Result<(), ProposalValidationError> {
+        let parent_header = self.parent;
+        if self.proposal.header.height() != parent_header.height() + 1 {
+            return Err(ProposalValidationError::InvalidHeight {
+                parent_height: parent_header.height(),
+                proposal_height: self.proposal.header.height(),
+            });
+        }
+        Ok(())
     }
+    /// Validate timestamp is not decreasing relative to parent and is
+    /// within a given tolerance of system time. Tolerance is
+    /// currently 12 seconds. This value may be moved to configuration
+    /// in the future.
+    fn validate_timestamp(&self) -> Result<(), ProposalValidationError> {
+        self.proposal
+            .validate_timestamp_non_dec(self.parent.timestamp())?;
 
-    let ValidatedState {
-        block_merkle_tree,
-        fee_merkle_tree,
-        ..
-    } = state;
+        // Validate timestamp hasn't drifted too much from system time.
+        // Do this check first so we don't add unnecessary drift.
```

Comment:
Nit: comment is now potentially misleading because in this function the drift check is done last.

Reply:
cea4a80a5beb971559ee661cc4af193233eb1e9f